### PR TITLE
create and handle level shifters

### PR DESCRIPTION
**Pull request by ahmadelrouby**
- Created ODB structure for UPF Level Shifters
- Added handler function for 'set_level_shifter' command in UPF to create and populate ODB with the appropriate data for level shifters
- Testing both update and create functionality

### CODE REVIEW

**Comment by QuantamHD** [comment]
Looks like your generated odb files are out of date. https://github.com/The-OpenROAD-Project/OpenROAD/actions/runs/6868196432/job/18678191742?pr=4265

Try running the generate script again.

File: src/upf/src/upf.cpp [hunk 1]
```diff
@@ -1017,4 +1017,181 @@ bool eval_upf(sta::dbNetwork* network, utl::Logger* logger, odb::dbBlock* block)
...
+    if (pd == nullptr) {
+      logger->warn(utl::UPF,
+                   38,
+                   "Couldn't find power domain {} for level shifter {}",
+                   domain,
+                   name);
+      return false;
+    }
+    ls = static_cast<odb::dbLevelShifter*>(
+        odb::dbLevelShifter::create(block, name, pd));
```
Comment (maliberty) [inline src/upf/src/upf.cpp:1064]: why is a cast needed here?

File: src/upf/src/upf.cpp [hunk 2]
```diff
@@ -1017,4 +1017,181 @@ bool eval_upf(sta::dbNetwork* network, utl::Logger* logger, odb::dbBlock* block)
...
+    ls->setNoShift(std::stoi(no_shift));
+  }
+
+  if (strlen(force_shift) > 0) {
+    ls->setForceShift(std::stoi(force_shift));
+  }
+
+  ls->setUseEquivalence(strcmp(use_equivalence, "FALSE") ? true : false);
+  ls->setUseFunctionalEquivalence(
+      strcmp(use_functional_equivalence, "FALSE") ? true : false);
```
Comment (maliberty) [inline src/upf/src/upf.cpp:1129]: you might use strcasecmp to make it case insensitive

File: src/upf/src/upf.cpp [hunk 3]
```diff
@@ -1017,4 +1017,181 @@ bool eval_upf(sta::dbNetwork* network, utl::Logger* logger, odb::dbBlock* block)
...
+    logger->warn(utl::UPF,
+                 41,
+                 "Couldn't find level shifter {} to exclude element {}",
+                 level_shifter_name,
+                 exclude_element);
+    return false;
+  }
+
+  std::string _exclude_element(exclude_element);
+  ls->addExcludeElement(_exclude_element);
```
Comment (maliberty) [inline src/upf/src/upf.cpp:1171]: ls->addExcludeElement(exclude_element);
should work fine.

File: src/upf/src/upf.cpp [hunk 4]
```diff
@@ -1017,4 +1017,181 @@ bool eval_upf(sta::dbNetwork* network, utl::Logger* logger, odb::dbBlock* block)
...
+  std::string _element(element);
+  ls->addElement(_element);
+
+  return true;
+}
+
+bool exclude_level_shifter_element(utl::Logger* logger,
+                                   odb::dbBlock* block,
+                                   char* level_shifter_name,
+                                   char* exclude_element)
```
Comment (maliberty) [inline src/upf/src/upf.cpp:1158]: const

File: src/odb/src/db/dbLevelShifter.cpp [hunk 5]
```diff
@@ -0,0 +1,537 @@
...
+  _block->_levelshifter_hash.insert(shifter);
+
+  domain->addLevelShifter((dbLevelShifter*) shifter);
+
+  return (dbLevelShifter*) shifter;
+}
+
+void dbLevelShifter::destroy(dbLevelShifter* shifter)
+{
+  // TODO
```
Comment (maliberty) [inline src/odb/src/db/dbLevelShifter.cpp:498]: ??

**Review by maliberty — CHANGES_REQUESTED** [review]

File: src/odb/src/db/dbLevelShifter.cpp (hunk 5 above)
Comment (ahmadelrouby) [reply src/odb/src/db/dbLevelShifter.cpp]: updated

**Comment by maliberty** [comment]
Please fix 
image

**Comment by maliberty** [comment]
You should also merge master to pick up #4291

**Comment by maliberty** [comment]
Please merge master and resolve the conflicts and this can go in.  Thanks for waiting.

**Comment by ahmadelrouby** [comment]
@maliberty Should be okay now. Just need your approval.

File: src/odb/src/db/dbLevelShifter.cpp [hunk 6]
```diff
@@ -0,0 +1,544 @@
...
+  obj->_force_shift = force_shift;
+}
+
+bool dbLevelShifter::isForceShift() const
+{
+  _dbLevelShifter* obj = (_dbLevelShifter*) this;
+  return obj->_force_shift;
+}
+
+void dbLevelShifter::setLocation(std::string location)
```
Comment (maliberty) [inline src/odb/src/db/dbLevelShifter.cpp:390]: Can the code generator be told to take these by const & ?

File: src/odb/src/db/dbPowerDomain.cpp [hunk 7]
```diff
@@ -154,6 +155,11 @@ dbIStream& operator>>(dbIStream& stream, _dbPowerDomain& obj)
   stream >> obj._x2;
   stream >> obj._y1;
   stream >> obj._y2;
+  // User Code Begin >>
+  if (stream.getDatabase()->isSchema(db_schema_level_shifter)) {
+    stream >> obj._levelshifters;
+  }
+  // User Code End >>
```
Comment (maliberty) [inline src/odb/src/db/dbPowerDomain.cpp:162]: Could this be replaced by specifying 'schema' on the _levelshifters field in the schema?

**Review by maliberty — APPROVED** [review]
We'll save the code gen for another PR.

**Comment by maliberty** [comment]
@fgaray FYI on another schema change

**Pull request merged**